posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,246 @@
+execute: 
+  cache: true
+---
+
+```{r, echo = FALSE}
+knitr::opts_chunk$set(
+  cache = TRUE,
+  message = FALSE
+)
+```
+
+R package developers are constantly faced with design decisions about user
+friendliness. Such design decisions often include checking that inputs are
+rightly specified and returning feedback early, deciding on how to deal with
+issues arising during code execution within a function, and how to
+structure outputs.
+
+These design decisions often result in taking on dependencies. Dependencies are
+external packages that enable a function to work properly. For example, the
+`tidyverse` is a collection of packages that are often used together to
+make data analysis easier. The `tidyverse` is a dependency for many
+packages that are built on top of it.
+
+Dependencies should only be adopted after carefully considering their pros and
+cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
+to guide the decision to take on dependencies that might lead to slower
+processes. In data analytics pipelines, speed is often a key consideration,
+especially when working with large datasets. Therefore, it is important to
+consider how the new dependency might impact the speed of an existing
+process.
+
+For this demonstration, we will look at condition signalling,
+that is, how to signal to a user that an issue has arisen during code
+execution [^2]. Condition signalling in R is classified into errors,
+warnings, and messages. Errors refer
+to cases where something fatal happens during code execution so that it halts,
+for example, when a function is run without specifying the required arguments.
+
+```{r}
+try(rnorm())
+```
+
+Warnings are used to handle less fatal issues with code execution that do not
+need to halt the process. For example, when you compare two vectors of
+different lengths, R will throw a warning but will still return a result.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Messages are used to provide useful information about processes. For
+example, when you load a package, R will throw a message to indicate that the
+package has been loaded.
+
+```{r}
+library(dplyr)
+```
+
+Base R ships with functionalities for condition signalling. There is
+`stop()` for signaling errors, `warning()` for throwing warnings, and
+`message()` for throwing messages. These base R
+functions are, however, not flexible enough for advanced
+formatting: colour, progress bar, contextual pluralisation, and so forth.
+
+Package developers often have to adopt external dependencies for flexibility
+in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
+that will not be discussed here, but of note is its `cli_abort()` and
+`cli_warn()` functions for signalling errors and warnings respectfully.
+
+`cli` is a great candidate for flexibility in condition signalling but one
+might naturally wonder if this comes at a cost. Here,
```

Comment:
Before reading this post, I would have naively assumed that {cli} was performant enough not to care about understanding/quantifying its speed. Therefore, it might be worth mentioning that benchmarking can be used to check whether a introduced function from an external package slows down the code even when it is not expected. This point about unexpected slowdowns/bottlenecks might be better placed in the conclusion.

Reply:
Yes, that makes sense. I've made the change here a2a5ddf1625a16ddbb03aacdf1ffae0dc55635a6.